tribits/core/package_arch/TribitsPackageDependencies.cmake

```diff
@@ -0,0 +1,204 @@
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# ************************************************************************
+# @HEADER
+
+
+################################################################################
+#
+# TribitPackageDependencies.cmake
+#
+# Contains macros and functions related to defining internal package and
+# external package/TPL dependencies with a little bit of logic.
+#
+################################################################################
+
+
+include_guard()
+
+include(TribitsParseArgumentsHelpers)
+include(MessageWrapper)
+
+cmake_policy(SET CMP0011 NEW) # include() does policy push and pop
+cmake_policy(SET CMP0057 NEW) # Support if ( ... IN_LIST ... )
+
+
+# @MACRO: tribits_external_package_define_dependencies()
+#
+# Macro called from inside of a ``FindTPL<tplName>Dependencies.cmake`` file to
+# define the direct upstream dependencies an external package/TPL.
+#
+# Usage::
+#
+#   tribits_external_package_define_dependencies( <tplName>
+#     DEPENDENCIES <upstreamTpl_0> <upstreamTpl_1>:<vis_1> ... )
+#
+# The listed upstream dependencies ``<upstreamTpl_i>`` are other external
+# packages/TPLs listed before this external packages/TPL ``<tplName>`` in a
+# `<repoDir>/TPLsList.cmake`_ file.  Each upstream dependency can include a
+# visibility specification ``<vis_i>`` that can be added to the dependency
+# using a colon ``:`` with ``<upstreamTpl_1>:<vis_1>`` where ``<vis_i>`` can
+# take the allowed values:
+#
+# * ``PUBLIC``: The include directories along with the libraries will be
+#   propagated downstream to clients of ``<tplName>``.
+#
+# * ``PRIVATE``: Only the libraries will be propagated downstream to clients
+#   of ``<tplName>``.
+#
+# If ``<vis_i>`` is not specified, then the default is ``PRIVATE``.  (If a
+# package needs the include directories from some external package/TPL, then
+# it should list that external package/TPL as a direct dependency and not
+# expect to get include directories from indirect dependencies.)
+#
+macro(tribits_external_package_define_dependencies
+    tplName
+  )
+
+  # Parse arguments
+  cmake_parse_arguments(
+     PARSE "" "" # prefix, options, one_value_keywords
+     "DEPENDENCIES"  # multi_value_keywords
+     ${ARGN}
+     )
+  tribits_check_for_unparsed_arguments(PARSE)
+  tribits_assert_parse_arg_one_or_more_values(PARSE  DEPENDENCIES)
+
+  set(libAllDependencies "")
+  foreach (upstreamTplDepEntry  IN  LISTS  PARSE_DEPENDENCIES)
+    list(APPEND libAllDependencies ${upstreamTplDepEntry})
```

Comment:
I think I was going to put in checks for the entries while looping over them.  But later I realized that I can't do that efficiently until later.  Also, I realized that when I combine the handling of external packages/TPLs and internal packages, the needed checks will come for free (because I already have them for internal packages).

I will change this to avoid the loop.

Reply:
Addressed in #498